lib/resque/worker.rb

```diff
@@ -226,33 +227,34 @@ def work(interval = 5.0, &block)
           working_on job
 
           procline "Processing #{job.queue} since #{Time.now.to_i} [#{job.payload_class_name}]"
-          if @child = fork(job)
-            srand # Reseeding
-            procline "Forked #{@child} at #{Time.now.to_i}"
+          if fork_per_job?
+            run_hook :before_fork, job
             begin
-              Process.waitpid(@child)
-            rescue SystemCallError
-              nil
+              @child = fork do
+                unregister_signal_handlers if term_child
+                perform(job, &block)
+                exit! unless run_at_exit_hooks
+              end
+            rescue NotImplementedError
+              @fork_per_job = false
             end
-            job.fail(DirtyExit.new("Child process received unhandled signal #{$?.stopsig}")) if $?.signaled?
-          else
-            unregister_signal_handlers if will_fork? && term_child
-            begin
-
-              reconnect if will_fork?
-              perform(job, &block)
-
-            rescue Exception => exception
-              report_failed_job(job,exception)
-            end
-
-            if will_fork?
-              run_at_exit_hooks ? exit : exit!
+            if @child
+              srand # Reseeding
+              procline "Forked #{@child} at #{Time.now.to_i}"
+              begin
+                Process.waitpid(@child)
+              rescue SystemCallError
+                nil
+              end
+              job.fail(DirtyExit.new("Child process received unhandled signal #{$?.stopsig}")) if $?.signaled?
+              @child = nil
             end
           end
+          unless fork_per_job?
```

Comment:
Yup, the extra check is for the fork fallback.  It looks like we are missing a test for that, since changing to an `else` doesn't cause any tests to fail.